### DISCUSSION TITLE
Write your own contaminants?

### DISCUSSION THREAD

**Discussion by detorrentel** [mwshinn/PyDDM#86] [Ask for help]
Similarly to how you can write your own Drift and noise class, I am wondering if there is a way to write our own contaminants/mixture models? It is in the bullet point but there is no link yet.
Let me explain what I am trying to do and maybe you even have ideas on how I could code it.
I am trying to model a process where I know that the first x seconds are more contaminated than after that. We model a "decision" where the people "make" more mistakes (not related to the interesting part) during those first x seconds.
I am using the poisson mixture and it works ok-ish but I want to rewrite it to have a higher rate of contamination at the beginning. Any idea how I could do that?
Thanks

**Comment by mwshinn** [comment]
Yes you can!  You need to make an Overlay object.  There is an example of something similar in the docs, but it is only for non-decision time (which is also implemented as an Overlay): https://pyddm.readthedocs.io/en/latest/cookbook/nondecision.html#gaussian-distributed-non-decision-time

Another example is OverlayExponentialRewardMixture from the following: https://pyddm.readthedocs.io/en/latest/cookbook/papers/shinn2020.html which changes the parameters of the exponential mixture model (PoissonMixture) based on a reward bias.  (From [Shinn et al (2020)](https://www.jneurosci.org/content/40/38/7326))  You can ignore the _test method and the decorators in this example.

Here is the implementation of the mixture model from the PyDDM source code that you can modify to suit your needs:

```python
class OverlayExponentialMixture(Overlay):
    name = "Poisson distribution mixture model (lapse rate)"
    required_parameters = ["pmixturecoef", "rate"]
    def apply(self, solution):
        assert self.pmixturecoef >= 0 and self.pmixturecoef <= 1
        choice_upper = solution.choice_upper
        choice_lower = solution.choice_lower
        m = solution.model
        cond = solution.conditions
        undec = solution.undec
        evolution = solution.evolution
        # To make this work with undecided probability, we need to
        # normalize by the sum of the decided density.  That way, this
        # function will never touch the undecided pieces.
        norm = np.sum(choice_upper)+np.sum(choice_lower)
        lapses = lambda t : 2*self.rate*np.exp(-1*self.rate*t)
        X = m.dt * np.arange(0, len(choice_upper))
        Y = lapses(X)
        Y /= np.sum(Y)
        choice_upper = choice_upper*(1-self.pmixturecoef) + .5*self.pmixturecoef*Y*norm # Assume numpy ndarrays, not lists
        choice_lower = choice_lower*(1-self.pmixturecoef) + .5*self.pmixturecoef*Y*norm
        return Solution(choice_upper, choice_lower, m, cond, undec, evolution)
```

The idea is that you take apart the PyDDM Solution object, change the distributions to your liking, and then build a replacement Solution object.
